helix-view/src/document.rs

```diff
@@ -291,8 +378,58 @@ impl Document {
         }
     }
 
+    pub(crate) async fn save_impl<'a>(
+        text: &'a Rope,
+        writer: &'a mut (impl tokio::io::AsyncWriteExt + Unpin),
+        encoder: &'a mut encoding_rs::Encoder,
+    ) -> Result<(), anyhow::Error> {
+        let iter = text.chunks();
+        let iter_len = iter.clone().count();
+        let mut buf = [0u8; BUF_SIZE];
+        let mut total_written = 0;
+
+        for (idx, chunk) in iter.enumerate() {
```

Comment:
Okay, so what we actually want to do is use a zero-sized final chunk as a sentinel that indicates we're at the end.  So instead of the enumerate, what we want to do is something like this:

```rust
for chunk in text.chunks().filter(|c| !c.is_empty()).chain(std::iter::once("")) {
```

(Untested: might have syntax errors.  And the filter is needed because Ropey doesn't guarantee non-zero-sized chunks.)

Then inside the loop, you just check if `chunk.is_empty()`, and if it is then that means you're at the end, and the last loop will act as a cleanup of any dangling data left in `encoder`'s internal buffer.